File: IKeep.Lib/Models/GenericTask.cs
﻿using IKeep.Lib.Core;
using IKeep.Lib.Models;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace IKeep.Lib.Models
{
    public class GenericTask : Entity
    {
        public string Ref { get; set; }
        public string Description { get; set; }
        public TimeSpan Duration { get; set; }
        public Period Period { get; set; }
        public Guid PriorityId { get; set; }
        [JsonIgnore]
        public virtual Priority Priority { get; set; }

        public string PriorityName
        {
            get
            {
                return Priority == null ? "" : Priority.Name;
            }
        }
        public Guid FormatId { get; set; }
        [JsonIgnore]
        public virtual Format Format { get; set; }

        public string FormatName
        {
            get
            {
                return Format == null ? "" : Format.Name;
            }
        }

        public Guid CategoryId { get; set; }
        [JsonIgnore]
        public virtual Category Category { get; set; }

        public string CategoryName
        {
            get
            {
                return Category == null ? "" : Category.Name;
            }
        }

        [JsonIgnore]
        public virtual ICollection<Task> Tasks { get; set; }
        public List<Guid> TasksIds
        {
            get
            {
                return Tasks == null ? new List<Guid>() : Tasks.Select(x => x.Id).ToList();
            }
        }

        [JsonIgnore]
        public virtual ICollection<GenericElementGenericTask> GenericElementGenericTasks { get; set; }
        public List<Guid> GenericElementGenericTasksIds
        {
            get
            {
                return GenericElementGenericTasks == null ? new List<Guid>() : GenericElementGenericTasks.Select(x => x.Id).ToList();
            }
        }
    }

    public enum Period
    {
        Daily,
        Weekly,
        Monthly,
        Bimonthly,
        Quarterly,
        Semester,
        Yearly,
        TwoYearly,
        FourYearly
    }
}

File: IKeep.Web/ClientApp/src/models/GenericTask.js
import Entity from './Entity'

class GenericTask extends Entity
{
    constructor(json)
    {
        super(json);
        if(json)
        {
            this.Ref = json.ref;
            this.Description = json.description;
            this.Duration = json.duration;
            this.Period = json.period;
            this.PriorityId = json.priorityId;
            this.FormatId = json.formatId;
            this.CategoryId = json.categoryId;
            this.TasksIds = json.tasksIds;

            this.PriorityName = json.priorityName;
            this.FormatName = json.formatName;
            this.CategoryName = json.categoryName;
        }
        else
        {
            this.Ref = "";
            this.Description = "";
            this.Duration = "";
            this.Period = Periodicity;
            this.PriorityId = "";
            this.FormatId = "";
            this.CategoryId = "";
            this.TasksIds = "";

            this.PriorityName = "";
            this.FormatName = "";
            this.CategoryName = "";            
        }
    }

}
export default GenericTask;

const Periodicity = Object.freeze({
    Daily: 0,
    Weekly: 1,
    Monthly: 2,
    Bimonthly: 3,
    Quarterly: 4,
    Semester: 5,
    Yearly: 6,
    TwoYearly: 7,
    FourYearly: 8
})

